alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java

```diff
@@ -4114,256 +2915,960 @@ public void reindexNodeByQuery(String query) throws IOException, AuthenticationE
+                TermQuery q = new TermQuery(new Term(fieldName, fieldType.readableToIndexed(Long.toString(id))));
+                TopDocs topDocs = searcher.search(q, 1);
+                return topDocs.totalHits > 0;
+            }
+            finally
+            {
+                ofNullable(refCounted).ifPresent(RefCounted::decref);
+            }
         }
+    }
 
-        public void collect(int doc) throws IOException
+    private void lock(Object id) throws IOException
+    {
+        long startTime = System.currentTimeMillis();
+        while(!lockRegistry.add(id))
         {
-            long txnId = currentLongs.get(doc);
-
-            //System.out.println("########### Floor Filter #############:"+doc+":"+txnId);
+            try
+            {
+                Thread.sleep(100);
+            }
+            catch (InterruptedException e)
+            {
+                // I don't think we are concerned with this exception.
+            }
 
-            if(txnId >= txnFloor)
+            if(System.currentTimeMillis() - startTime > 120000)
             {
-                this.leafDelegate.collect(doc);
+                throw new IOException("Unable to acquire lock on nodeId " + id + " after " + 120000 + " msecs.");
             }
         }
     }
 
-    class TxnCollector extends DelegatingCollector
+    private synchronized void unlock(Object id)
     {
+        lockRegistry.remove(id);
+    }
 
-        private NumericDocValues currentLongs;
-        private long txnFloor;
-        private long txnCeil;
-        private LongHashSet txnSet = new LongHashSet(1000);
-
+    private SolrDocumentList executeQueryRequest(SolrQueryRequest request, SolrQueryResponse response, SolrRequestHandler handler)
+    {
+        handler.handleRequest(request, response);
 
-        public TxnCollector(long txnFloor)
-        {
-            this.txnFloor = txnFloor;
-            this.txnCeil = txnFloor+500;
-        }
+        NamedList values = response.getValues();
+        return (SolrDocumentList) values.get(RESPONSE_DEFAULT_IDS);
+    }
 
-        public void doSetNextReader(LeafReaderContext context) throws IOException
+    /**
+     * Adds tenant information to an authority, <strong>if required</strong>, such that jbloggs for tenant example.com
+     * would become jbloggs@example.com
+     *
+     * @param authority   The authority to mutate, if it matches certain types.
+     * @param tenant      The tenant that will be added to the authority.
+     * @return The new authority information
+     */
+    private String addTenantToAuthority(String authority, String tenant)
+    {
+        switch (AuthorityType.getAuthorityType(authority))
         {
-            currentLongs = context.reader().getNumericDocValues(FIELD_INTXID);
+            case GROUP:
+            case EVERYONE:
+            case GUEST:
+                if (tenant.length() != 0)
+                {
+                    return authority + "@" + tenant;
+                }
```

Comment:
We should probably have a break statement here.  Would this switch statement be simpler as an if statement?